lib/AST/ASTContext.cpp

```diff
@@ -11182,3 +11182,46 @@ APFixedPoint ASTContext::getFixedPointMin(QualType Ty) const {
   assert(Ty->isFixedPointType());
   return APFixedPoint::getMin(getFixedPointSemantics(Ty));
 }
+
+RecordDecl *ASTContext::getCachedTypeApp(const RecordDecl *Base, ArrayRef<const Type *> TypeArgs) {
+  auto it = CachedTypeApps.find(std::make_pair(Base, TypeArgs));
+  if (it == CachedTypeApps.end()) return nullptr;
+  return it->second;
+}
+
+void ASTContext::addCachedTypeApp(const RecordDecl *Base, ArrayRef<const Type *> TypeArgs, RecordDecl *Inst) {
+  assert(Base == Base->getCanonicalDecl() && "Expected key to be canonical decl");
+  assert(Base == Inst->baseDecl() && "Base decl must match in key and value");
+  assert((getCachedTypeApp(Base, TypeArgs) == nullptr) && "Type application is already cached");
+  CachedTypeApps.insert(std::make_pair(std::make_pair(Base, TypeArgs), Inst));
+}
+
+ArrayRef<RecordDecl *> ASTContext::getDelayedTypeApps(RecordDecl *Base) {
+  assert(Base->isGeneric() && "Expected a generic struct");
+  assert(Base->getCanonicalDecl() == Base && "Key should be a canonical decl");
+  auto Iter = DelayedTypeApps.find(Base);
+  if (Iter == DelayedTypeApps.end()) return ArrayRef<RecordDecl *>();
+  return ArrayRef<RecordDecl *>(*Iter->second);
+}
+
+void ASTContext::addDelayedTypeApp(RecordDecl *TypeApp) {
+  assert(TypeApp->isInstantiated() && TypeApp->isDelayedTypeApp() && "Expected a delayed type application");
+  auto Base = TypeApp->baseDecl();
+  assert(Base->getCanonicalDecl() == Base && "Base should be a canonical decl");
+  llvm::SmallVector<RecordDecl *, 4> *Delayed = nullptr;
+  auto Iter = DelayedTypeApps.find(Base);
+  if (Iter == DelayedTypeApps.end()) {
+    Delayed = new (*this) llvm::SmallVector<RecordDecl *, 4>();
```

Comment:
Clarification: Who deletes the memory for "Delayed"? Can we get rid of the "new" for "SmallVector"?

Reply:
There's a region of memory associated with the AST Context.  The data until the AST Context is deallocated..  This is using placement-new to put the vector in that context.  

However, as you point out, it may be advisable to not allocate the memory in the first place.  SmallVector is designed to only heap allocate if the specified number of elements (4) is exceeded.  I believe the definition of DelayedTypesApps needs to be changed.

Reply:
Done